### ISSUE TITLE
VSCode Sticky Scroll no longer works in ERB files after 0.7.10

### ISSUE THREAD

**Issue by brianclinkenbeard** [Shopify/ruby-lsp#2314]
### Reproduction steps

1. Start Ruby LSP > 0.7.10 using vscode 
2. Open an ERB file
3. Regression occurs where sticky scroll no longer appears

Example with a YAML file:
On 0.7.10:
![image](https://github.com/user-attachments/assets/bace9252-c03b-4c4e-9c2d-245135a3b6cf)
On 0.7.12:
![image](https://github.com/user-attachments/assets/afcbfb82-a2c3-41ef-bbab-cc490be8350b)

Likely related to the ERB code navigation change in #2235. I am using v0.17.7 of the language server.

**Comment by andyw8** [comment]
Settings beginning with `ruby.` are not used by Ruby LSP – those are for a different extension.

**Comment by st0012** [comment]
@brianclinkenbeard I can't reproduce the "fix" of downgrading extension. I do notice that I got sticky scroll back briefly when the extension restarts, like after downgrading or upgrading it.
Can you let me know if you still get the sticky scroll with the following steps:

- Downgrade to `v0.7.10` and restart the extension
- Close the file
- Open the file again

Additionally, can you check the language mark (example in screenshot) always indicates `erb`?

<img width="10%" alt="Screenshot 2024-07-18 at 21 14 25" src="https://github.com/user-attachments/assets/d178c9f4-3ee4-4373-bc77-c070b557e750">

And are you sure `v0.7.10` is when the behaviour first changed?

**Comment by brianclinkenbeard** [comment]
@st0012 Yup, following those steps maintains sticky scroll on the .yml.erb file I have open. The language mark in the bottom right always indicates `erb` (except when no version of Ruby LSP is installed, then it goes to `Ruby`). It seems that `v0.7.11` is the first version when this occurs, and the sticky scroll goes away right after the LSP server starts.

**Comment by brianclinkenbeard** [comment]
It looks like `v0.7.14` fixes this issue, but I'm not sure why: I didn't disable the new `rubyLsp.erbSupport` setting, which defaults to true. :thinking:

**Comment by vinistock** [comment]
Since the issue seems to be gone, I'm closing this until we receive other reports. If you experience the issue again, let us know!

**Comment by brianclinkenbeard** [comment]
@vinistock Can we reopen this issue, actually? It turned out that this being "fixed" previously was from an issue where Ruby LSP was failing to start and enable the erb syntax.

**Comment by brianclinkenbeard** [comment]
This appears to be related to the `documentSymbols` feature. When this feature turned off, sticky scroll and the erb formatting works correctly.

**Comment by vinistock** [comment]
@brianclinkenbeard do you have a reproduction example so that we can investigate? If this is related to document symbols in ERB files, it might be some specific code pattern that triggers it.

**Comment by brianclinkenbeard** [comment]
Hi @vinistock, I've spent some time narrowing this down to these reproduction steps:
1. Make sure `documentSymbols` feature is enabled to reproduce the issue (and LSP has restarted if necessary)
2. View a `.yml.erb` file following this pattern in VSCode:
```yml
<% CONST = nil %>

list:
  nested_list:
    - item
    - item
    - item
    # ...

list2: # For whatever reason, a trailing list is needed to ever show sticky scroll
```
Disabling documentSymbols will restore the sticky scroll.

**Comment by vinistock** [comment]
I did a bit of an investigation on this. The problem is that we are not merging full document request results in delegate scenarios.

[Here](https://github.com/Shopify/ruby-lsp/blob/dadc26a98faee30cabf2ae034ba58ea884f19811/vscode/src/client.ts#L578), we delegate requests like completion, go to definition and signature help to the host language's service.

Unfortunately, that solution doesn't work for full document requests, such as document symbol or semantic highlighting. Those are significantly more complicated because you'd need to:

- Run the request for the Ruby part through the Ruby LSP
- Also delegate the request to the host language's service
- Correctly combine the results (which may not be trivial, especially in the case of semantic highlighting when tokens are expressed as a delta relative to the previous token)

We could always delegate document symbols, but that means that the Ruby symbols would be missing from ERB files.

Not a super simple problem to solve, it would be good to understand what other language services with delegation are doing. For example, does the JavaScript language service merge semantic tokens when mixed with HTML or CSS?
